docs/dev_101_services.md

```diff
@@ -197,3 +197,62 @@ If you need more control over the service call, you can also pass an async funct
 async def custom_set_sleep_timer(entity, service_call):
     await entity.set_sleep_timer(service_call.data['sleep_time'])
 ```
+
+## Response Data
+
+Services may optionally respond to a service call with data for powering more advanced automations. The use of response data for services is meant for
+cases that do not fit the Home Assistant state. For example, a response stream of repeated objects, or API would be better queried on demand or filtered such as a search.
+
+Example code:
+
+```python
+import datetime
+import voluptuous as vol
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, ServiceCall, ServiceResponse
+from homeassistant.helpers import config_validation as cv, entity_platform, service
+
+
+SEARCH_ITEMS_SERVICE_NAME = "search_items"
+SEARCH_ITEMS_SCHEMA = vol.Schema({
+    vol.Required("start"): datetime.datetime,
+    vol.Required("end"): datetime.datetime,
+})
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the platform."""
+
+    async def search_items(call: ServiceCall) -> ServiceResponse:
+        """Search in the date range and return the matching items."""
+        if not call.return_response:
+            raise ValueError("Request did not ask for response data")
+        items = await my_client.search(call.data["start"], call.data["end"])
+        return {
+            "items": [
+                {
+                    "summary": item["summary"],
+                    "description": item["description"],
+                } for item in items
+            ],
+        }
+
+      hass.services.async_register(
+          DOMAIN,
+          SEARCH_ITEMS_SERVICE_NAME,
+          search_items,
+          schema=SEARCH_ITEMS_SCHEMA,
+          supports_response=True,
+      )
+```
+
+There are some additional implementation standards:
+
+- You must set `supports_response` when registering a service so that the core
+knows which services support response data.
+- All response data should be serializable in json `JsonObjectType`. This is so that it can interoperate with other parts of the system such as the frontend.
+- You may conditionally check the `ServiceCall` property `return_response` to
+decide whether or not response data should be returned (e.g. if it is large or
+expensive to fetch). When not set, the service may return `None`.
+- Errors must be raised as exceptions, such as `HomeAssistantError`. The response data is not allowed to contain error codes or statuses which are used for error handling. We want to keep error handling simple to avoid error handling mistakes.
+- Response data should not be used if there is a simpler alternative provided by the state or entity model.
```

Comment:
Remove this, or at least link and/or describe the simpler approaches.

Reply:
Thanks, I've clarified this. Basically i'm trying to say something like: don't use this for a temperature sensor when you should use a temperature sensor.